### ISSUE TITLE
WebAssembly maximum memory 2GB causes Out of Memory error on iOS Safari 16.2

### ISSUE THREAD

**Issue by TyounanMOTI** [godotengine/godot#70621]
### Godot version

4.0.dev (f382a2b)

### System information

iOS 16.2 Safari, iPhone 13 Pro

### Issue description

When exporting for Web and running on iOS 16.2 Safari, out of memory error occurs at WebAssembly.Memory initialization.
This error comes because `maximum` argument is 2048MB which is quite bigger for iOS.
When I changed `WASM_MEM_MAX` to 256MB for example at platform/web/detect.py: 197, the error disappeared.

```Python
    env.Append(LINKFLAGS=["-s", "WASM_MEM_MAX=256MB"])
```

### Steps to reproduce

1. Build Godot Web export with option `module_raycast_enabled=no` to disable WebAssembly SIMD usage which seems to be not yet supported on iOS Safari 16.2.
2. Export minimal reproduction project to Web.
3. Out of memory error log will be found on JavaScript console (reproduced).
4. Change WASM_MEM_MAX to 256MB and rebuild Web export.
5. Re-export minimal reproduction project.
6. Rotating unlit white cube is correctly rendered (reproduced).

### Minimal reproduction project

[simple_rotating_white_cube.zip](https://github.com/godotengine/godot/files/10306671/simple_rotating_white_cube.zip)

**Comment by Calinou** [comment]
cc @Faless

**Comment by TyounanMOTI** [comment]
Even if doing these workaround, sometimes page crash occurs while loading indicator. Maybe related to #67949.

**Comment by ChildLearningClub** [comment]
I'm getting the same error, posted a screen shot for reference. I think this is related to #67949 as well. 
![IMG_5210](https://user-images.githubusercontent.com/25701774/217950882-b8d4102d-2a97-4c18-8951-edeb4fdc01bd.JPG)

**Comment by stefano-elysium** [comment]
@TyounanMOTI I tried following your steps and I still could not get my project to run. 
I receive an error, "Unhandled Promise Rejection: CompileError: WebAssembly.Module doesn't parse at byte 5: can't get Function local's type in group 1, in function at index 34516"
Could you please upload your export template?

**Comment by stefano-elysium** [comment]
I will try recompiling with module_raycast_enabled=no, but just as a fyi I tested with the module I have and was unable to run on ios16.4 (although I couldn't figure out how to check if the browser updated too)

**Comment by michaldev** [comment]
> ![image](https://user-images.githubusercontent.com/85214725/228463016-32e52af7-370d-4c13-a558-bb0db5f00ae9.png) With module_raycast_enabled=no I managed to run it!

Is any method for use this solution without compiling engine?

**Comment by dyegoaurelio** [comment]
Unfortunately, at least from my testing, the only realiable way to have web exports that run on most platforms is to use godot 3

**Comment by stefano-elysium** [comment]
@dyegoaurelio the issue is the same if you attempt to run godot 3 with multithreading on.

**Comment by dyegoaurelio** [comment]
> @dyegoaurelio the issue is the same if you attempt to run godot 3 with multithreading on.

The support for godot 3 web exports is better than godot 4 even with multithreading on (which is an option feature)

For multithreading to work, the browser just needs to support [Shared Array Buffer](https://caniuse.com/sharedarraybuffer)

**Comment by stefano-elysium** [comment]
Have you tried running a multithreaded godot 3 version on the iphone? I receive the same issue as with godot 4

**Comment by Calinou** [comment]
@akien-mga Should we use `module_raycast_enabled=no` for official web builds? It's only used for CPU-based occlusion culling, which I gather isn't going to be used often in web projects (as most of them are 2D or simple 3D scenes).

**Comment by akien-mga** [comment]
Yeah we can probably disable the raycast module by default on Web, directly in `detect.py`.

I hadn't really considered the memory footprint from including everything that _can_ compile, so we might want to review other optional features to see if some would make sense to disable by default if we think that they're very unlikely to be used for Web projects. This would need to be documented of course as some users can be taken aback when a feature that works on PC seems broken/missing on Web.

**Comment by ChronoDK** [comment]
In my testing WASM_MEM_MAX=256MB and module_raycast_enabled=no does allow Godot 4 games to start on iOS Safari. They seem to crash after a while though, but this could easily be a problem with Safari and not Godot. At least the games will start now :-)

**Comment by michaldev** [comment]
> In my testing WASM_MEM_MAX=256MB and module_raycast_enabled=no does allow Godot 4 games to start on iOS Safari. They seem to crash after a while though, but this could easily be a problem with Safari and not Godot. At least the games will start now :-)

Some players have reported similar issues on Chrome with Android.

**Comment by Ansraer** [comment]
@akien-mga I think github accidentally closed this issue, could you reopen it? This issue was originally about Godot starting with a too high web assembly memory requirement, which hasn't been fixed by the linked PR.

While disabling the ray cast module due to the lack of simd support in wasm is also necessary, it isn't a full solution to this problem.

**Comment by michaldev** [comment]
On Godot 4.2 dev 6 still doesn't work.

https://github.com/godotengine/godot/assets/5132385/a9c141ef-c259-4541-9d85-35638bc35c5c

**Comment by ChronoDK** [comment]
That is as far as I have gotten as well. With 256MB it loads and looks fine for a few seconds, then it crashes.

**Comment by Faless** [comment]
> That is as far as I have gotten as well. With 256MB it loads and looks fine for a few seconds, then it crashes.

Should probably open an issue upstream: https://bugs.webkit.org/

**Comment by kisg** [comment]
Did someone already reproduce the issue with the upstream Webkit? This is how to build it from source: https://trac.webkit.org/wiki/BuildingGtk#BuildingWebKitGTKfromgit

If it could be reproduced there, it would be easier to debug, both for us and for the WebKit developers.

**Comment by teiman** [comment]
Monkey testing stuff around I found a trick.

`
    wasmMemory = new WebAssembly.Memory({
       "initial": INITIAL_MEMORY / 65536,
       //"maximum": 2147483648 / 65536,
       "maximum": 268000000 / 65536,//Tei: It was 2048 MB, but iOS choke on it
       "shared": true
      });
`

**Comment by geekley** [comment]
I'm getting this issue on Firefox on Android 10 as well, on 4.3.stable. It doesn't always happen, but sometimes it's stuck on 100% in the loading screen, and the browser console (visible by connecting to Desktop Firefox USB debugger) logs an "internal error out of memory" after a long time.

So, there's 2 issues. One is the memory issue itself. Second is the fact no feedback is shown to the user, so they could be waiting forever without knowing any sort of error happened. There should be some sort of error message if this out-of-memory issue happens.

**Comment by adamscott** [comment]
I wasn't able to reproduce the issue/MRP on my iPad Pro M4. But maybe it's because it's an iPad Pro M4.

**Comment by barthy-koeln** [comment]
Leaving breadcrumbs for other devs as I found myself coming back to this issue a dozen times over the last weeks.
___

People experiencing the OOM errors described here should first rule out that it's this known memory leak:

Any out-of-memory issues in Safari (iOS and macOS) for my project were resolved with #107948 (included in 4.5-beta2 and up).

**Comment by akien-mga** [comment]
Can anyone confirm whether @barthy-koeln's comment that this may have been resolved by #107948 holds true in your use case?
